File: app/src/main/java/fmrsabino/moviesdb/ui/base/PresenterLoader.java
package fmrsabino.moviesdb.ui.base;

import android.content.Context;
import android.support.v4.content.Loader;

import rx.functions.Func0;

public class PresenterLoader<T extends Presenter> extends Loader<T> {
    private T presenter;
    private final Func0<T> creator;

    public PresenterLoader(Context context, Func0<T> creator) {
        super(context);
        this.creator = creator;
    }

    @Override
    protected void onStartLoading() {
        if (presenter != null) {
            deliverResult(presenter);
            return;
        }
        forceLoad();
    }

    @Override
    protected void onForceLoad() {
        presenter = creator.call();
        deliverResult(presenter);
    }

    @Override
    protected void onReset() {
        presenter.onDestroyed();
        presenter = null;
    }
}

File: app/src/main/java/fmrsabino/moviesdb/ui/base/PresenterLoader.kt
package fmrsabino.moviesdb.ui.base

import android.content.Context
import android.support.v4.content.Loader

class PresenterLoader<T : BasePresenter<*>>(context: Context, private val creator: () -> T) : Loader<T>(context) {
    private var presenter: T? = null

    override fun onStartLoading() {
        presenter?.let { deliverResult(presenter); return }
        forceLoad()
    }

    override fun onForceLoad() {
        presenter = creator()
        deliverResult(presenter)
    }

    override fun onReset() {
        presenter?.onDestroyed()
        presenter = null
    }
}
